File: src/commands/account-allocations.ts
import { getRunEnv, invokeTauri, invokeWeb, logger, RUN_ENV } from "@/adapters";
import type { AccountAllocation, NewAccountAllocation, UpdateAccountAllocation } from "@/lib/types";

export const listAccountAllocations = async (
  virtualAccountId: string,
): Promise<AccountAllocation[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("list_account_allocations", { virtualAccountId });
      case RUN_ENV.WEB:
        return invokeWeb("list_account_allocations", { virtualAccountId });
      default:
        throw new Error("Unsupported");
    }
  } catch (e) {
    logger.error("Error listing allocations.");
    throw e;
  }
};

export const createAccountAllocation = async (
  allocation: NewAccountAllocation,
): Promise<AccountAllocation> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("create_account_allocation", { allocation });
      case RUN_ENV.WEB:
        return invokeWeb("create_account_allocation", { allocation });
      default:
        throw new Error("Unsupported");
    }
  } catch (e) {
    logger.error("Error creating allocation.");
    throw e;
  }
};

export const updateAccountAllocation = async (
  allocationId: string,
  changes: UpdateAccountAllocation,
): Promise<AccountAllocation> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("update_account_allocation", { allocationId, changes });
      case RUN_ENV.WEB:
        return invokeWeb("update_account_allocation", { allocationId, changes });
      default:
        throw new Error("Unsupported");
    }
  } catch (e) {
    logger.error("Error updating allocation.");
    throw e;
  }
};

export async function deleteAccountAllocation(allocationId: string): Promise<void> {
  switch (getRunEnv()) {
    case RUN_ENV.DESKTOP:
      return invokeTauri("delete_account_allocation", { allocationId });
    case RUN_ENV.WEB:
      return invokeWeb("delete_account_allocation", { allocationId });
    default:
      throw new Error("Unsupported");
  }
}

File: src-tauri/src/commands/account_allocations.rs
use std::sync::Arc;

use crate::{
    context::ServiceContext,
    events::{emit_resource_changed, ResourceEventPayload},
};
use log::{debug, error};
use tauri::{State};
use serde_json::json;
use wealthfolio_core::accounts::{AccountAllocation, NewAccountAllocation, UpdateAccountAllocation};

async fn recalc_virtual_and_emit(
    state: &Arc<ServiceContext>,
    handle: &tauri::AppHandle,
    virtual_account_id: &str,
) -> Result<(), String> {
    state
        .snapshot_service()
        .recalculate_virtual_account(virtual_account_id)
        .await
        .map_err(|e| format!("Failed to recalc virtual snapshots: {}", e))?;

    emit_resource_changed(
        handle,
        ResourceEventPayload::new(
            "virtual_account",
            "snapshots_updated",
            serde_json::json!({ "account_id": virtual_account_id }),
        ),
    );

    Ok(())
}

#[tauri::command]
pub async fn create_account_allocation(
    allocation: NewAccountAllocation,
    state: State<'_, Arc<ServiceContext>>,
    handle: tauri::AppHandle,
) -> Result<AccountAllocation, String> {
    debug!("Adding new account allocation...");

    let virtual_account_id = allocation.virtual_account_id.clone();

    let result = state
        .account_allocation_service()
        .create_allocation(allocation)
        .await;

    match result {
        Ok(acc) => {
            // ✅ recompute virtual snapshots immediately
            recalc_virtual_and_emit(&state, &handle, &virtual_account_id).await?;

            emit_resource_changed(
                &handle,
                ResourceEventPayload::new(
                    "account",
                    "created",
                    json!({ "account_id": acc.id }),
                ),
            );
            Ok(acc)
        }
        Err(e) => {
            error!("Failed to add new account allocation: {}", e);
            Err(format!("Failed to add new account allocation: {}", e))
        }
    }
}

#[tauri::command]
pub async fn list_account_allocations(
    virtual_account_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<AccountAllocation>, String> {
    debug!("Listing allocations for virtual_account_id={}", virtual_account_id);

    state
        .account_allocation_service()
        .list_for_virtual_account(&virtual_account_id)
        .map_err(|e| {
            error!("Failed to list allocations: {}", e);
            format!("Failed to list allocations: {}", e)
        })
}

#[tauri::command]
pub async fn update_account_allocation(
    allocation_id: String,
    changes: UpdateAccountAllocation,
    state: State<'_, Arc<ServiceContext>>,
    handle: tauri::AppHandle,
) -> Result<AccountAllocation, String> {
    debug!("Updating allocation allocation_id={}", allocation_id);

    let result = state
        .account_allocation_service()
        .update_allocation(&allocation_id, changes)
        .await;

    match result {
        Ok(acc) => {
            // ✅ recompute
            recalc_virtual_and_emit(&state, &handle, &acc.virtual_account_id).await?;

            emit_resource_changed(
                &handle,
                ResourceEventPayload::new(
                    "account",
                    "updated",
                    json!({ "account_id": acc.id }),
                ),
            );
            Ok(acc)
        }
        Err(e) => {
            error!("Failed to update allocation: {}", e);
            Err(format!("Failed to update allocation: {}", e))
        }
    }
}

#[tauri::command]
pub async fn delete_account_allocation(
    allocation_id: String,
    state: State<'_, Arc<ServiceContext>>,
    handle: tauri::AppHandle,
) -> Result<(), String> {
    let deleted = state
        .account_allocation_service()
        .delete_allocation(&allocation_id)
        .await
        .map_err(|e| format!("Failed to delete allocation: {}", e))?;

    recalc_virtual_and_emit(&state, &handle, &deleted.virtual_account_id).await?;
    Ok(())
}

